### ISSUE TITLE
SFT vs Continuous Pretraining

### ISSUE THREAD

**Issue by ghtaro** [huggingface/trl#556]
Hi,

I have a question on example scripts on SFT below.
 https://github.com/lvwerra/trl/blob/170d58ffcede84b3bc822294317fc2bb6df85865/examples/research_projects/stack_llama/scripts/supervised_finetuning.py#L87

I was thinking that SFT is trained on a set of pairs ("input sequence" and "output sequence") by minimizing a -loglikihood of "output sequence" conditioned on "input sequence".

However, If I am not wrong, the above code does minimizing a -loglikihood of "input + output sequence". It is more like (continuous) pretraining (I guess that is why we called SFT instead of continuous pretraining....).

I was able to run the script with my own dataset, but I feel I have not understood fully what I am doing here.

If I am missing something, it would be very helpful to spot it out.

**Comment by younesbelkada** [comment]
Hi @ghtaro 
Your understanding is correct, in that scenario it sounds more like a continuous training; and it seems to perform as good as "conditioned" training in practice from the things I have seen around (for e.g. Guanaco model)
Since the architecture is a decoder model the simplest way to reproduce "conditional" fine-tuning would be to mask out the tokens of the input sequence during the computation of the loss function. This can be performed using the `DataCollatorForCompletionOnlyLM` that you can use directly from `trl` : https://github.com/lvwerra/trl/blob/main/trl/trainer/utils.py#L55

**Comment by ghtaro** [comment]
Hi @younesbelkada 

Thank you very much for your reply.
I understood the difference and the reason why you use the one for continuous training. 

My use case would be for llm to generate just the answer for any user input question (I mean I do not require llm to generate question), so to me it would be reasonable to try to use DataCollatorForCompletionOnlyLM (at least for loss convergence).

**Comment by younesbelkada** [comment]
Perfect, let us know if you face into any issue while using that collator!

**Comment by younesbelkada** [comment]
FYI, here is the section on the docs on how to FT on completions only using `DataCollatorForCompletionOnlyLM`: https://huggingface.co/docs/trl/main/en/sft_trainer#train-on-completions-only

**Comment by ghtaro** [comment]
@younesbelkada thank you very much! For some reasons, I could not run `trl/examples/research_projects/stack_llama/scripts/supervised_finetuning.py` for 
- Falcon 7B
- single node with multi-gpu

Instead I was able to run yours (https://gist.github.com/younesbelkada/f48af54c74ba6a39a7ae4fd777e72fe8) by adding DataCollatorForCompletionOnlyLM without errors.

Thank you again for your support.

**Comment by lvwerra** [comment]
@ghtaro can you share the issue you had with the script?
